examples/step-74/step-laplace-FEIV.cc

```diff
@@ -0,0 +1,1190 @@
+      : Function<dim>()
+      , test_case(test_case)
+    {}
+    virtual void           value_list(const std::vector<Point<dim>> &points,
+                                      std::vector<double> &          values,
+                                      const unsigned int             component = 0) const;
+    virtual Tensor<1, dim> gradient(const Point<dim> & point,
+                                    const unsigned int component = 0) const;
+
+  private:
+    Functions::LSingularityFunction ref;
+    Test_Case                       test_case;
+  };
+
+  template <int dim>
+  void Solution<dim>::value_list(const std::vector<Point<dim>> &points,
+                                 std::vector<double> &          values,
+                                 const unsigned int             component) const
+  {
+    switch (this->test_case)
+      {
+        case convergence_rate:
+          {
+            using numbers::PI;
+            for (unsigned int i = 0; i < values.size(); ++i)
+              values[i] = std::sin(2 * PI * points[i][0]) *
+                          std::sin(2 * PI * points[i][1]);
+            break;
+          }
+        case l_singularity:
+          {
+            for (unsigned int i = 0; i < values.size(); ++i)
+              values[i] = ref.value(points[i]);
+            break;
+          }
+        default:
+          {
+            Assert(false, ExcNotImplemented());
+          }
+      }
+  }
+
+
+  template <int dim>
+  Tensor<1, dim> Solution<dim>::gradient(const Point<dim> & point,
+                                         const unsigned int component) const
+  {
+    switch (this->test_case)
+      {
+        case convergence_rate:
+          {
+            Tensor<1, dim> return_value;
+            using numbers::PI;
+            return_value[0] = 2 * PI * std::cos(2 * PI * point[0]) *
+                              std::sin(2 * PI * point[1]);
+            return_value[1] = 2 * PI * std::sin(2 * PI * point[0]) *
+                              std::cos(2 * PI * point[1]);
+            return return_value;
+            break;
+          }
+        case l_singularity:
+          {
+            return ref.gradient(point);
+            break;
+          }
+        default:
+          {
+            Assert(false, ExcNotImplemented());
+          }
+      }
+  }
+
+  template <int dim>
+  class RHS : public Function<dim>
+  {
+  public:
+    RHS(Test_Case test_case)
+      : Function<dim>()
+      , test_case(test_case)
+    {}
+    virtual void value_list(const std::vector<Point<dim>> &points,
+                            std::vector<double> &          values,
+                            const unsigned int             component = 0) const
+    {
+      switch (test_case)
+        {
+          case convergence_rate:
+            {
+              using numbers::PI;
+              for (unsigned int i = 0; i < values.size(); ++i)
+                values[i] = 8 * PI * PI * std::sin(2 * PI * points[i][0]) *
```

Comment:
here, and in all other places: make sure to denote numbers as doubles: ``8.`` instead of ``8``.